wip - streaming api for RPC service

yusefnapora opened the pull request:
So far untested and missing some important features (like catchup of previous blocks / events).  Compiles tho :)

yusefnapora committed:
sketch rpc update streaming api

yusefnapora committed:
use ref to last block instead of offset, add journalBlockPublished event type

yusefnapora committed:
basic streaming events - no catchup yet

parkan commented on protocol/src/main/protobuf/Transactor.proto:
```diff
@@ -14,14 +14,34 @@ message UpdateRequest {
 
     rpc UpdateChain(UpdateRequest) returns
-        (MultihashReference);
+        (ChainUpdateDescription);
```
oh yeah 😅

parkan replied:
I'm not sure I love the "switch two words around and add either Request or Description (and maybe drop a word)" method of deriving these names

parkan replied:
I kind of want to just say `UpdateChainResult`

vyzo replied:
why are we changing this return type?

yusefnapora replied:
yeah, this name kind of sucks.  I changed the return type to more closely match the ChainEntry, where you have a ref to the canonical and the cell instead of just the cell.  I figure it could be useful in a client, because you might only care about updates to a subset of all canonicals, so having the canonical ref could be handy.  `UpdateChainResult` sounds good tho

vyzo replied:
The client already has the canonical in order to do the update, so it doesn't need it.
What could be perhaps useful is the chainPrevious in the ChainEntry.

parkan commented on transactor/src/main/scala/io/mediachain/copycat/TransactorService.scala:
```diff
@@ -3,21 +3,30 @@ package io.mediachain.copycat
-import io.mediachain.protocol.Transactor.JournalError
-import io.mediachain.protocol.transactor.Transactor.{InsertRequest, TransactorServiceGrpc, UpdateRequest}
+import io.mediachain.protocol.Transactor.{JournalError, JournalListener}
+import io.mediachain.protocol.transactor.Transactor.{MultihashReference => _, _}
```
are we trying to avoid shadowing another MultihashReference? also yusef you keep discovering new and strange ways of aliasing imports, good job

yusefnapora replied:
I don't know how that happened, lol 😛  Maybe intellij has become self aware.

parkan commented on protocol/src/main/protobuf/Transactor.proto:
```diff
@@ -14,14 +14,34 @@ message UpdateRequest {
     rpc LookupChain(MultihashReference) returns
         (MultihashReference);
+
+    rpc StreamJournalUpdates(JournalStreamRequest) returns
```
`rpc JournalStream(JournalStreamRequest)` ?

parkan commented on transactor/src/main/scala/io/mediachain/copycat/TransactorService.scala:
```diff
@@ -104,19 +113,92 @@ class TransactorService(client: Client,
+    // per-observer, etc.
+
+    journalEventObservers.synchronized {
+      journalEventObservers.add(responseObserver)
```
paging @bigs since I think he researched this

parkan commented on transactor/src/main/scala/io/mediachain/copycat/TransactorService.scala:
```diff
@@ -104,19 +113,92 @@ class TransactorService(client: Client,
+    if (state == ClientState.Disconnected) {
+      journalEventObservers.synchronized {
+        journalEventObservers.foreach { observer =>
+          observer.onError(
```
why did they call this `onError`? it's not a listener! 😞

yusefnapora committed:
revert return type change for updateChain rpc call, rename chain update event proto to UpdateChainResult

vyzo commented on protocol/src/main/protobuf/Transactor.proto:
```diff
@@ -14,6 +14,24 @@ message UpdateRequest {
+message UpdateChainResult {
+    MultihashReference canonicalRef = 1;
+    MultihashReference chainHeadRef = 2;
+    MultihashReference chainPreviousHeadRef = 3;
```
is there anything gained by calling things XXXRef instead of just XXX?

yusefnapora replied:
no, I think you're right.  using Ref is inconsistent with the other protos as well

yusefnapora replied:
How do you feel about `canonical`, `chain`, and `chainPrevious`?  It's closer to the ChainEntry naming but I think using `canonical` instead of `ref` is a bit clearer.   Maybe it's better to be consistent with ChainEntry though.

vyzo replied:
that sounds much better.

yusefnapora committed:
naming things

- rename StreamJournalUpdates -> JournalStream
- rename members of UpdateChainResult

yusefnapora committed:
remove cancelled journal stream observers

bigs committed:
Add simple helper fns for development

bigs committed:
Factor out copycat listener into separate class

bigs committed:
Merge remote-tracking branch 'origin/yn-rpc-streaming' into yn-rpc-streaming

Conflicts:
	transactor/src/main/scala/io/mediachain/copycat/TransactorService.scala

bigs committed:
Create `addObservers` fn for listener

vyzo commented on transactor/src/main/scala/io/mediachain/transactor/Entrypoints.scala:
```diff
@@ -0,0 +1,40 @@
+    val datastoreConfig = DynamoDatastore.Config(
+      "Mediachain",
+      new BasicAWSCredentials("", ""),
+      Some("http://localhost:8000")
```
why is this useful at all?
The servers are running in the cluster with a server program, you can already start a server by running JournalServer and the transactor server by running RpcService.
Also, it's hardcoded to local dynamo and uses a table we are not currently using (we are still on test).
Also the db name in dynamo comes from your access key, and there is none here which is bad.

vyzo commented:
@bigs the EntryPoints thing is useless I think. 
We arelady have a program to run the server, no need to have a hardcoded mess in the source tree. 
Please kill it!

vyzo commented on protocol/src/main/protobuf/Transactor.proto:
```diff
@@ -14,6 +14,24 @@ message UpdateRequest {
+        MultihashReference canonicalInserted = 2;
+        UpdateChainResult chainUpdated = 3;
+        MultihashReference journalBlockPublished = 4;
+    }
```
naming-wise, can we drop the past tense from the events?

bigs committed:
Upgrade cats, add dogs

bigs committed:
Add offset to update stream request

bigs committed:
Implement `toStream` for streaming blocks in chain

bigs committed:
Add basic backfilling support for transactor svc

Pending:
- Buffering of new blocks while replaying old blocks
- Working off queue of new blocks after replaying is done

bigs committed:
cats => dogs in translation_engine

bigs committed:
Implement queuing of events while backfilling

bigs committed:
Prune unnecessary Promise import

bigs committed:
Clean up stream traversal w/ collect

bigs committed:
Use a stack to ensure proper order of backfilling

vyzo commented:
belay the comment about toStream -- I just read the diff and thought it was in Datastore!

yusefnapora committed:
bump scala-multihash commmit

yusefnapora committed:
mediate tensions between cats and dogs

- dogs pulls in cats 0.4.1, so depending on cats 0.6.0 was causing
  problems

yusefnapora committed:
use JournalStreamRequest.lastJournalBlock instead of offset, add logging

yusefnapora committed:
playback journal entries for catchup blocks

parkan commented on protocol/src/main/scala/io/mediachain/protocol/Datastore.scala:
```diff
@@ -354,5 +359,19 @@ object Datastore {
+
+    def toStream(datastore: Datastore): Streaming[JournalBlock] = {
+      val next: Eval[Streaming[JournalBlock]] = chain match {
+        case Some(ref) => Eval.always {
```
is the point of `Eval` here to avoid blowing the stack?

bigs replied:
`Eval.always` means "always execute this block" instead of caching the result

parkan replied:
right, I get what `Eval` generally does, but given that the two branches here are `.always` and a `.now` returning a constant, I'm assuming it's used as a trampoline for the recursion.

parkan commented:
@bigs let's address https://github.com/mediachain/mediachain/commit/c124ad791ec068245fedcb1f45094e4f51abb33b#commitcomment-17783823 and I'll merge this.

Also agree with https://github.com/mediachain/mediachain/pull/117#discussion_r66030131 but that can go on top

parkan commented:
@bigs let's address https://github.com/mediachain/mediachain/commit/c124ad791ec068245fedcb1f45094e4f51abb33b#commitcomment-17783823 and merge?

bigs commented:
@parkan that entire file can be removed from the project. it's been superseded by the main function yusef and i wrote

yusefnapora commented:
I thought @bigs meant to remove the `Entrypoints`, not the `RpcService`.   I think we still need `RpcService` to run the facade.  But `Entrypoints` is just for easier repl development.

parkan commented:
Oops yeah I misread that one sec

parkan commented:
Actually both RpcService and Entrypoints hardcode the DynamoDatastore as local, what's the deal? @bigs

Commit:
Merge branch 'master' into yn-rpc-streaming

Commit:
Remove Endpoints, use new DynamoDatastore.Config
